### PR TITLE
Implements the LineInititialMapper strategy

### PR DESCRIPTION
**Pull request by ammareltigani**
This PR implements the main initial mapping strategy highlighted in http://tinyurl.com/cirq-qubit-routing

### CODE REVIEW

File: cirq-core/cirq/transformers/__init__.py [hunk 1]
```diff
@@ -44,7 +44,13 @@
     two_qubit_gate_product_tabulation,
 )
 
-from cirq.transformers.routing import AbstractInitialMapper, HardCodedInitialMapper, MappingManager
+
+from cirq.transformers.routing import (
+    MappingManager,
+    HardCodedInitialMapper,
+    AbstractInitialMapper,
+    LineInitialMapper,
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/__init__.py:52]: nit: Order lexicographically.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Concrete implementation of AbstractInitialMapper that places lines of qubits onto the device."""
+
+from typing import Dict, Optional, TYPE_CHECKING
+import networkx as nx
+
+from cirq import value, _compat
+from cirq.transformers import routing
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:21]: Don't import `routing` since it can lead to circular imports (routing imports `line_initial_mapper` and `line_initial_mapper.py` imports `routing`). 
Import the specific submodule instead (`from cirq.transformers.routing import initial_mapper`)

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        if nx.is_directed(device_graph):
+            self.device_graph = nx.DiGraph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(sorted(list(device_graph.edges)))
+        else:
+            self.device_graph = nx.Graph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(
+                sorted(list(sorted(edge) for edge in device_graph.edges))
+            )
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:48]: We are doing this mainly to ensure that `_value_equality_values_` can return a tuple which has a property that equal graphs have equal tuples. However, we really don't need any other property of `_value_equality_values_` (like automatically generated hash function?) so I'd suggest that we do the following:
1. Remove `_value_equality_values_` and directly implement an `__eq__` method that uses `nx.utils.graphs_equal(self.device_graph, other.device_graph)` for checking graph equality. 
2. Not do this preprocessing in every class that needs to store a device graph.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+from cirq import value, _compat
+from cirq.transformers import routing
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class LineInitialMapper(routing.AbstractInitialMapper):
+    """Places logical qubits in the circuit onto physical qubits on the device."""
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:29]: Update docstring to reflect that specific strategy used by the `LineInitialMapper`.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+            self.device_graph = nx.Graph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(
+                sorted(list(sorted(edge) for edge in device_graph.edges))
+            )
+
+    def _make_circuit_graph(self, circuit: 'cirq.AbstractCircuit') -> nx.Graph:
+        """Creates a (potentially incomplete) qubit connectivity graph of the circuit.
+
+        Iterates over the moments circuit from left to right drawing edges between logical qubits
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:53]:
```suggestion
        Iterates over moments in the circuit from left to right and adds edges between logical qubits
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        that:
+            (1) have degree < 2, and
+            (2) that are involved in a 2-qubit operation in the current moment.
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:56]:
```suggestion
            (2) are involved in a 2-qubit operation in the current moment.
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+        At this point the graph is forest of paths and/or simple cycles. For each simple cycle, make
+        it a path by removing the last edge that was added to it.
+
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            The (potentially incomplete) qubit connectivity graph of the circuit.
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:64]:
```suggestion
            The (potentially incomplete) qubit connectivity graph of the circuit, which is guaranteed to be a forest of line graphs.
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+        """
+        circuit_graph = nx.Graph()
+        edge_order = 0
+        for op in circuit.all_operations():
+            circuit_graph.add_nodes_from(op.qubits)
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:69]: Add all `circuit.all_qubits()` as nodes to the graph before starting the for-loop. Adding it for each operation is wasteful and might be incorrect (what if there's a qubit that doesn't have any operation acting on it?).
```suggestion
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+        while found:
+            try:
+                cycle = nx.find_cycle(circuit_graph)
+                edge_to_remove = max(
+                    cycle, key=lambda x: circuit_graph.edges[x[0], x[1]]['edge_order']
+                )
+                circuit_graph.remove_edge(*edge_to_remove)
+            except nx.exception.NetworkXNoCycle:
+                found = False
+        return circuit_graph
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:85]: Let's use union find to make this cleaner and more efficient. 
```suggestion
    circuit_graph = nx.Graph()
    circuit_graph.add_nodes_from(circuit.all_qubits())
    union_find = nx.utils.UnionFind(elements=circuit.all_qubits())
    for op in circuit.all_operations():
        if (
            len(op.qubits) == 2
            and all(d < 2 for _, d in circuit_graph.degree(op.qubits))
            and union_find[op.qubits[0]] != union_find[op.qubits[1]]
        ):
            union_find.union(*op.qubits)
            circuit_graph.add_edge(*op.qubits)
    return circuit_graph
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        return self._initial_mapping(circuit.freeze())
+
+    @_compat.cached_method
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:109]: Caching an entire (large) circuit would be expensive. In what scenario do we believe there's an added advantage of caching? Can we just cache the circuit graph instead of caching the entire circuit? Would two different circuits with same circuit graph have the same initial mapping?

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Left a first round of comments. Still reviewing the `_initial_mapping` method.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(sorted(list(device_graph.edges)))
+        else:
+            self.device_graph = nx.Graph()
+            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
+            self.device_graph.add_edges_from(
+                sorted(list(sorted(edge) for edge in device_graph.edges))
+            )
+
+    def _make_circuit_graph(self, circuit: 'cirq.AbstractCircuit') -> nx.Graph:
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:50]: Consider returning the graph as a `List[List[cirq.Qid]]` (i.e. a list of paths) instead of an `nx.Graph`; since that's what's needed in the `initial_mapping` function. It'll help avoid conversion to/from networkx representation to paths representation. It would also give you the "sort by largest connected component" by just sorting the outer list based on size of inner list and "iterating from one end point to other" would just correspond to iterating on the inner list. This would also get rid of the `next_logical` method.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        def next_logical(current_logical: 'cirq.Qid') -> Optional['cirq.Qid']:
+            for neighbor in circuit_graph.neighbors(current_logical):
+                if circuit_graph.nodes[neighbor]["mapped"] is False:
+                    return neighbor
+            return None
+
+        for pq in self.device_graph.nodes:
+            self.device_graph.nodes[pq]["mapped"] = False
+        for lq in circuit_graph.nodes:
+            circuit_graph.nodes[lq]["mapped"] = False
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:151]: I don't think we need a `mapped` attribute on `circuit_graph` nodes.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+            return self._closest_unmapped_qubit(physical_center)
+
+        def next_logical(current_logical: 'cirq.Qid') -> Optional['cirq.Qid']:
+            for neighbor in circuit_graph.neighbors(current_logical):
+                if circuit_graph.nodes[neighbor]["mapped"] is False:
+                    return neighbor
+            return None
+
+        for pq in self.device_graph.nodes:
+            self.device_graph.nodes[pq]["mapped"] = False
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:149]: Can we use a set `mapped = {}` instead, and query "if current_physical in mapped" instead of `self.device_graph.nodes[current_physical]["mapped"] is False` ? 

This would help not increase the memory footprint of the entire class (by adding attributes to the `self.device_graph` instead of using a local data structure that gets destroyed after going out of scope).

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+    """Places logical qubits in the circuit onto physical qubits on the device."""
+
+    def __init__(self, device_graph: nx.Graph) -> None:
+        """Initializes a LineInitialMapper.
+
+        Args:
+            device_graph: device graph
+        """
+        # TODO: Tanuj, should this logic be done instead at the beginning of the routing transformer
+        # so that we don't have to repeat it for each InitialMapper and for the MappingManager?
```
Comment (mpharrigan) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:38]: open a github issue and link it in any TODO comments

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py (hunk 3 above)
Comment (mpharrigan) [reply cirq-core/cirq/transformers/routing/line_initial_mapper.py:48]: @ammareltigani why do we need equality support?

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+            device (values).
+        """
+        qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {}
+        circuit_graph = self._make_circuit_graph(circuit)
+        physical_center = nx.center(self.device_graph)[0]
+
+        def next_physical(current_physical: 'cirq.Qid') -> 'cirq.Qid':
+            # use current physical if last logical line ended before mapping to it.
+            if self.device_graph.nodes[current_physical]["mapped"] is False:
+                return current_physical
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:129]: I think this line can be removed if we remove `current_physical = next_physical(current_physical)` from Line 160

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+            )
+
+    def _make_circuit_graph(self, circuit: 'cirq.AbstractCircuit') -> nx.Graph:
+        """Creates a (potentially incomplete) qubit connectivity graph of the circuit.
+
+        Iterates over the moments circuit from left to right drawing edges between logical qubits
+        that:
+            (1) have degree < 2, and
+            (2) that are involved in a 2-qubit operation in the current moment.
+        At this point the graph is forest of paths and/or simple cycles. For each simple cycle, make
```
Comment (mpharrigan) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:57]:
```suggestion
        At this point the graph is a forest of paths and/or simple cycles. For each simple cycle, make
```

**Review by tanujkhattar — CHANGES_REQUESTED** [review]
Few more comments, mainly around cleaning up the code in `self._initial_mapping`

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+        found = True
+        while found:
+            try:
+                cycle = nx.find_cycle(circuit_graph)
+                edge_to_remove = max(
+                    cycle, key=lambda x: circuit_graph.edges[x[0], x[1]]['edge_order']
+                )
+                circuit_graph.remove_edge(*edge_to_remove)
+            except nx.exception.NetworkXNoCycle:
+                found = False
```
Comment (mpharrigan) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:84]: you can probably(?) change this to `while True` and use `break` in this exception handler

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py (hunk 14 above)
Comment (ammareltigani) [reply cirq-core/cirq/transformers/routing/line_initial_mapper.py:38]: I removed the TODO here but I'll do that for the next time!

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py (hunk 4 above)
Comment (ammareltigani) [reply cirq-core/cirq/transformers/routing/line_initial_mapper.py:29]: I updated the docstring in `initial_mapping()`. Should I move it to the docstring for the class or keep it there?

**Comment by ammareltigani** [comment]
> @ammareltigani why do we need equality support?
I don't think we actually need caching nor equality support. 

We don't need caching because the user can easily save the mapping locally since we supply it as a simple dictionary. 

We don't need equality since equality would be (class name, device graph). Class name is easy to keep track of from the user's perspective and the device graph is something that is given by the user as well. So they equality can easily be tracked by the user if they desire to do so.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+            circuit_graph.nodes[lq]["mapped"] = False
+
+        current_physical = physical_center
+        for logical_cc in nx.connected_components(circuit_graph):
+            if len(logical_cc) == 1:
+                continue
+            # logical_cc is a set, make it a sorted list to guarantee deterministic behavior
+            logical_cc = sorted(logical_cc)
+
+            current_physical = next_physical(current_physical)
```
[the remark this answers is not in the document]
Comment (ammareltigani) [reply cirq-core/cirq/transformers/routing/line_initial_mapper.py:160]: The line ` if current_logical is not None:` tests for an edge case where if mapping n qubits on an n-qubit device we should not call next_physical when finished mapping the last logical qubit else will raise an error.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 19]
```diff
@@ -14,18 +14,18 @@
 
 """Concrete implementation of AbstractInitialMapper that places lines of qubits onto the device."""
 
-from typing import Dict, Optional, TYPE_CHECKING
+from audioop import reverse
```
Comment (mpharrigan) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:17]: this seems wrong

**Review by mpharrigan — APPROVED** [review]
no blockers from me

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

**Comment by ammareltigani** [comment]
@tanujkhattar A few notes:
1) The indeterminism issue stemming from different edge ordering can be resolved by sorting all the edges of the incoming device graph in the constructor in the same way as 'MappingManager'. Given this it might make sense to make the class a value_equality since we have the sorted graph anyway. This obviously introduces some overhead in the `__init__` method so another solution would be to sort the nodes given by each iteration of `bfs_successors()` in the `_find_closest_unmapped_qubit()` helper. Not sure what is more costly in this tradeoff.

2) There isn't an exact guarantee that is placement strategy will always make the 2-qubit gates in the first two timesteps executable. The exceptions are 2-qubit gates with qubits incident to an edge that was removed in cycle breaking. Here is a quote from https://arxiv.org/pdf/1902.08091.pdf : _. This
ensures that most of the gates in the first two timesteps can be applied without any swaps; the
only exceptions are those gates corresponding to the edges removed when breaking rings._

**Comment by tanujkhattar** [comment]
> 1. The indeterminism issue stemming from different edge ordering can be resolved by sorting all the edges of the incoming device graph in the constructor in the same way as 'MappingManager'. Given this it might make sense to make the class a value_equality since we have the sorted graph anyway. This obviously introduces some overhead in the `__init__` method so another solution would be to sort the nodes given by each iteration of `bfs_successors()` in the `_find_closest_unmapped_qubit()` helper. Not sure what is more costly in this tradeoff.

IIUC, the indeterminism is in graph construction in the tests then? Anyways, sorting once in the constructor to get rid of sorting in each bfs_successors and get_next_physical_qubit sounds good to me.

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+                else iter(circuit_graph[c1])
+            )
+            for q in c1_order:
+                if circuit_graph[c0][0] == q0:
+                    circuit_graph[c0].appendleft(q)
+                else:
+                    circuit_graph[c0].append(q)
+                component_id[q] = c0
+
+        return sorted([circuit_graph[c] for c in set(component_id.values())], key=len, reverse=True)
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:140]: nit: 
```suggestion
    circuit_graph: List[Deque['cirq.Qid']] = [deque([q]) for q in sorted(circuit.all_qubits())]
    component_id: Dict['cirq.Qid', int] = {q[0]: i for i, q in enumerate(circuit_graph)}
    partners: Dict['cirq.Qid', 'cirq.Qid'] = {}

    def pos_in_deque(q: 'cirq.Qid') -> Optional[int]:
        for i in [0, -1]:
            if circuit_graph[component_id[q]][i] == q:
                return i
        return None

    for op in circuit.all_operations():
        if protocols.num_qubits(op) != 2:
            continue
        q0, q1 = op.qubits
        # Keep track of partners for mapping isolated qubits later.
        partners[q0] = partners[q0] if q0 in partners else q1
        partners[q1] = partners[q1] if q1 in partners else q0
        # Find component IDs and positions in the paths.
        c0, c1 = component_id[q0], component_id[q1]
        pos0, pos1 = pos_in_deque(q0), pos_in_deque(q1)
        if pos0 is None or pos1 is None or c0 != c1:
            continue
        # Make sure c0/q0 are for the largest component.
        if len(circuit_graph[c0]) < len(circuit_graph[c1]):
            c0, q0, pos0, c1, q1, pos1 = c1, q1, pos1, c0, q0, pos0
        # copy smaller component into larger one.
        c1_iter = reversed(circuit_graph[c1]) if pos1 == -1 else iter(circuit_graph[c1])
        for q in c1_iter:
            circuit_graph[c0].insert(pos0, q)
            component_id[q] = c0
    graph = sorted([circuit_graph[c] for c in set(component_id.values())], key=len, reverse=True)
    return graph, partners```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+        Args:
+            circuit: the input circuit with logical qubits
+
+        Returns:
+            a dictionary that maps logical qubits in the circuit (keys) to physical qubits on the
+            device (values).
+        """
+        qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {}
+        circuit_graph = self._make_circuit_graph(circuit)
+        self.center = nx.center(self.device_graph)[0]
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:154]:
```suggestion
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+
+        def next_physical(current_physical: 'cirq.Qid') -> 'cirq.Qid':
+            # Greedily map to highest degree neighbor that that is available
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:157]:
```suggestion
            # Greedily map to highest degree neighbor that is available
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 23]
```diff
@@ -0,0 +1,221 @@
...
+            sorted_neighbors = sorted(
+                self.device_graph.neighbors(current_physical),
+                key=lambda x: self.device_graph.degree(x),
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:160]: Should we be sorting based on "unused degree", i.e. `x` > `y` if `x` has more unmapped neighbors than y?

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 24]
```diff
@@ -0,0 +1,221 @@
...
+                reverse=True,
+            )
+            for neighbor in sorted_neighbors:
+                if neighbor not in self.mapped_physicals:
+                    return neighbor
+            # If cannot map onto one long line of physical qubits, then break down into multiple
+            # small lines by finding nearest available qubit to the physical center
+            return self._closest_unmapped_qubit(self.center)
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:168]: Should we find a closest unmapped qubit to `current_physical` instead? If the outer loop of logical qubits is iterating over a connected line, then this would ensure that we find the closest node to the previous neighbour in line (i.e. `current_physical`), instead of forgetting about the line position and restarting from center?

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 25]
```diff
@@ -0,0 +1,221 @@
...
+            if len(logical_line) == 1:
+                first_isolated_idx = idx
+                break
+
+            for lq in logical_line:
+                self.mapped_physicals.add(pq)
+                qubit_map[lq] = pq
+                # Edge case: if mapping n qubits on an n-qubit device should not call next_physical
+                # when finished mapping the last logical qubit else will raise an error
+                if len(circuit.all_qubits()) != len(self.mapped_physicals):
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:182]: `len(circuit.all_qubits()) ` is very slow! `circuit.all_qubits()` iterates over all operations in the circuit, and adds all qubits to a frozen set and returns that frozen set, without any caching. This is because `cirq.Circuit` is a mutable object and number of qubits in the circuit can change with time. 

Let's precompute the number of qubits at the beginning of the function once, using `sum(len(x) for x in circuit_graph)`, and re-use the value here.

**Review by tanujkhattar — CHANGES_REQUESTED** [review]

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py (hunk 23 above)
Comment (ammareltigani) [reply cirq-core/cirq/transformers/routing/line_initial_mapper.py]: I'll keep note of this change and try it during benchmarking!

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py (hunk 24 above)
Comment (ammareltigani) [reply cirq-core/cirq/transformers/routing/line_initial_mapper.py]: I'll keep note of this change and try it during benchmarking!

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 26]
```diff
@@ -0,0 +1,223 @@
...
+            device (values).
+        """
+        mapped_physicals: Set['cirq.Qid'] = set()
+        qubit_map: Dict['cirq.Qid', 'cirq.Qid'] = {}
+        circuit_graph, partners = self._make_circuit_graph(circuit)
+
+        def next_physical(
+            current_physical: 'cirq.Qid', partner: 'cirq.Qid', isolated: bool = False
+        ) -> 'cirq.Qid':
+            # Greedily map to highest degree neighbor that is available
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:160]:
```suggestion
            if current_physical not in mapped_physicals:
                return current_physical

```

File: cirq-core/cirq/transformers/routing/line_initial_mapper_test.py [hunk 27]
```diff
@@ -0,0 +1,216 @@
...
+    )
+    cirq.testing.assert_same_circuits(circuit.transform_qubits(mapping), expected_circuit)
+
+
+@pytest.mark.parametrize(
+    "qubits, n_moments, op_density, random_state",
+    [
+        (5 * size, 20 * size, density, seed)
+        for size in range(1, 3)
+        for seed in range(10)
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper_test.py:152]:
```suggestion
        for seed in range(3)
```

File: cirq-core/cirq/transformers/routing/line_initial_mapper.py [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+                    pq = next_physical(pq, self.center)
+
+                # Edge case: if mapping n qubits on an n-qubit device should not call next_physical
+                # when finished mapping the last logical qubit else will raise an error.
+                elif i < len(circuit_graph) - 1:
+                    if len(circuit_graph[i + 1]) > 1:
+                        pq = next_physical(pq, self.center)
+                    else:
+                        partner = qubit_map[partners[lq]] if lq in partners else self.center
+                        pq = next_physical(pq, partner, isolated=True)
```
Comment (tanujkhattar) [inline cirq-core/cirq/transformers/routing/line_initial_mapper.py:190]: Let's clean this up to call `next_physical` n times instead of n+1 times to avoid edge cases.

**Review by tanujkhattar — APPROVED** [review]
LGTM % nits
